### PR TITLE
[Bugfix] Disable unintended CFG for Z-Image Turbo

### PR DESCRIPTION
**Pull request by Xenoryn**
## Purpose

Fix Z-Image-Turbo unexpectedly running classifier-free guidance (CFG) when the request omits `guidance_scale` or sets it to `0.0`.

The shared request parser normalizes a false-like guidance scale (including `0.0`) to `1.0` and leaves `guidance_scale_provided` unset. `ZImagePipeline` previously consumed the normalized value directly and enabled CFG for any value greater than zero. This caused two text encoder calls and doubled the DiT batch, making inference significantly slower than the expected no-CFG path.

This PR resolves the value inside `ZImagePipeline`:

- preserve an explicitly provided positive `guidance_scale`; and
- otherwise use Z-Image's no-CFG default, `0.0`.

The fix is model-scoped and does not change shared request parsing or other diffusion pipelines.

## Test Plan

Start the profiled server and send the same request before and after the fix:

```bash
vllm serve Tongyi-MAI/Z-Image-Turbo --omni \
  --port 30000 --diffusion-attention-backend FLASH_ATTN \
  --enable-diffusion-pipeline-profiler --log-stats
```

```bash
curl -sS http://127.0.0.1:30000/v1/images/generations \
  -H 'Content-Type: application/json' \
  -d '{
    "model": "Tongyi-MAI/Z-Image-Turbo",
    "prompt": "一只戴着宇航员头盔的橘猫，坐在月球表面，电影级光影",
    "size": "512x512",
    "n": 1,
    "num_inference_steps": 8,
    "guidance_scale": 0.0,
    "seed": 42,
    "response_format": "b64_json"
  }'
```

Also verify that an explicitly provided positive guidance scale still enables CFG.

## Test Result

Single-request A/B measurement with the same model, prompt, resolution, steps, seed, and attention backend:

| Metric | Before | After | Improvement |
| --- | ---: | ---: | ---: |
| Text encoder calls | 2 | 1 | Negative branch removed |
| Denoising latency | 57.115 ms/step | 33.527 ms/step | 1.70x |
| Pipeline forward | 434.091 ms | 244.601 ms | 1.77x |
| End-to-end latency | 458.771 ms | 270.151 ms | 1.70x |

- Python compilation: passed
- `git diff --check`: passed

**vLLM Version:** `0.24.0`

**vLLM-Omni Commit:** `c1d94d88bf2e03af84e6e172c5073d6da5608d5d`

**Hardware / Software:** NVIDIA H800, driver `570.148.08`; `FLASH_ATTN` backend.

### CODE REVIEW

File: vllm_omni/diffusion/models/z_image/pipeline_z_image.py [hunk 1]
```diff
@@ -168,6 +169,13 @@ class ZImagePipeline(nn.Module, DiffusionPipelineProfilerMixin, SupportsComponen
     _encoder_modules: ClassVar[list[str]] = ["text_encoder"]
     _vae_modules: ClassVar[list[str]] = ["vae"]
 
+    @staticmethod
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/z_image/pipeline_z_image.py]: Do other models have this same issue? Might be worth pulling it into a common func.

File: vllm_omni/diffusion/models/z_image/pipeline_z_image.py [hunk 2]
```diff
@@ -168,6 +169,13 @@ class ZImagePipeline(nn.Module, DiffusionPipelineProfilerMixin, SupportsComponen
...
+    def _resolve_guidance_scale(sampling_params: OmniDiffusionSamplingParams) -> float:
```
Comment (david6666666) [inline vllm_omni/diffusion/models/z_image/pipeline_z_image.py]: Please add a small regression test for omitted, `0.0`, and positive `guidance_scale`. This is easy to re-break because the shared request layer normalizes the value before Z-Image sees it.

File: vllm_omni/diffusion/models/z_image/pipeline_z_image.py (hunk 1 above)
Comment (Xenoryn) [reply vllm_omni/diffusion/models/z_image/pipeline_z_image.py]: @lishunyang12  Good point. I checked the other diffusion pipelines and found that several already use ``guidance_scale_provided`` to preserve their model-specific defaults after request normalization.  

I extracted the common logic into ``vllm_omni/diffusion/utils/param_utils.py::resolve_guidance_scale`` and updated Z-Image to use it with ``default=0.0``.

Should I migrate the other pipelines here too, or leave that for a follow-up?

Addressed in c95ce52a.

File: vllm_omni/diffusion/models/z_image/pipeline_z_image.py (hunk 2 above)
Comment (Xenoryn) [reply vllm_omni/diffusion/models/z_image/pipeline_z_image.py]: @david6666666 Added a parametrized regression test covering omitted ``guidance_scale``, explicit ``0.0``, and a positive value (``3.5``). 

Validation:
``python3 -m pytest -s -v `tests/diffusion/test_diffusion_request.py -m "core_model and diffusion and cpu"``
Result: 
All 9 tests pass.

Addressed in c95ce52a.

File: vllm_omni/diffusion/models/z_image/pipeline_z_image.py [hunk 3]
```diff
@@ -457,7 +458,7 @@ def forward(self, req: DiffusionRequestBatch) -> DiffusionOutput:
         generator = req.sampling_params.generator
         sigmas = req.sampling_params.sigmas
         max_sequence_length = req.sampling_params.max_sequence_length or 512
-        guidance_scale = req.sampling_params.guidance_scale
+        guidance_scale = resolve_guidance_scale(req.sampling_params, default=0.0)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/z_image/pipeline_z_image.py:461]: `do_classifier_free_guidance` is `_guidance_scale > 0`, so an explicit `guidance_scale=1.0` is still truthy → `provided=True` → CFG on with scale 1.0 (`2*pos - neg`). On a Turbo/distilled model a user passing the neutral 1.0 probably means no guidance. Gate on `> 1` instead of `> 0`?

File: vllm_omni/diffusion/utils/param_utils.py [hunk 4]
```diff
@@ -6,6 +6,22 @@
...
+) -> float:
+    """Resolve an explicit guidance scale or use a model-specific default.
+
+    ``OmniDiffusionRequest`` normalizes omitted and false-like scales before a
+    pipeline sees them, so consumers must use ``guidance_scale_provided`` to
+    distinguish an explicit positive value from the normalized sentinel.
+    """
+    if sampling_params.guidance_scale_provided:
+        return float(sampling_params.guidance_scale)
+    return float(default)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/utils/param_utils.py:22]: This only covers the primary scale — `guidance_scale_2` is still auto-filled to the normalized `1.0` in `request.py` with no `guidance_scale_2_provided` flag, so dual-guidance pipelines (wan2_2, omnigen2, nextstep) keep leaking the sentinel. Not z-image's problem, but the fix reads as complete when it's scoped to one scale.

File: tests/diffusion/test_diffusion_request.py [hunk 5]
```diff
@@ -72,3 +73,28 @@ def test_tp_seed_same_across_ranks_and_varies_across_requests():
...
     # Seeds must vary across requests (non-determinism preserved).
     assert len(set(seeds)) == n_requests, f"Expected {n_requests} unique seeds but got {len(set(seeds))}: {seeds}"
+
+
+@pytest.mark.parametrize(
+    ("sampling_kwargs", "normalized_scale", "scale_provided", "expected_scale"),
+    [
+        pytest.param({}, 1.0, False, 0.0, id="omitted"),
+        pytest.param({"guidance_scale": 0.0}, 1.0, False, 0.0, id="explicit-zero"),
+        pytest.param({"guidance_scale": 3.5}, 3.5, True, 3.5, id="explicit-positive"),
```
Comment (lishunyang12) [inline tests/diffusion/test_diffusion_request.py:83]: Add an `explicit-one` case (`{"guidance_scale": 1.0}` → provided `True`, expected `1.0`). That's the exact boundary this PR is about — 1.0 is truthy so it survives normalization and re-enables CFG.

**Review by lishunyang12 — COMMENTED** [review]
i think the fix is incomplete — an explicit `guidance_scale=1.0` is truthy so CFG still re-enables on Turbo. gating on `>1` would cover it.

**Comment by Xenoryn** [comment]
Hi @lishunyang12, sorry for the delayed response—I missed the email notification for your review.

Thank you for pointing out the `guidance_scale=1.0` boundary and the similar issue with `guidance_scale_2`. After revisiting the PR, I think these are valid concerns, but they also reveal a limitation of the approach I chose.

The original goal of this PR was to preserve explicit user input. Fully extending the current approach, however, would require introducing and maintaining additional model-specific defaults. For example, the same `ZImagePipeline` supports models with different expected defaults:

- `Tongyi-MAI/Z-Image-Turbo`: `guidance_scale=0.0`
- `Tongyi-MAI/Z-Image`: `guidance_scale=4.0`

This adds complexity and starts making model-specific choices on behalf of users, which is not what this PR originally intended to do.

I noticed that #4999 addresses the original sentinel problem more cleanly by distinguishing an omitted `guidance_scale` from an explicitly provided value. It preserves user input without requiring the additional model-default logic introduced by this PR, so I think it is a better solution to the original problem.

The `guidance_scale=1.0` CFG boundary and the corresponding `guidance_scale_2` behavior may still be worth addressing separately, but they seem broader than the original scope of this PR.

Given this, would you be comfortable with closing this PR in favor of #4999 and handling any remaining boundary or secondary-scale issues in focused follow-ups?
